app/controllers/mixins/breadcrumbs_mixin.rb

```diff
@@ -33,33 +34,51 @@ def data_for_breadcrumbs
           breadcrumbs.push(:title => @title)
         end
       else
+        x_node_text_set if action_name == 'tree_select' # get text of the active node
+
         # Append breadcrumb from title of the accordion (eg "Policies")
-        breadcrumbs.push(:title => accord_title, :key => "#{accord_name}_accord", :action => "accordion_select") if accord_title
+        breadcrumbs.push(:title => accord_title, :key => "root") if accord_title
+
+        # On special page (tagitems, politems, etc.)
+        if @tagitems || @politems || @ownershipitems || @retireitems
+          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems, true)) unless options[:hide_special_item]
+          breadcrumbs.push(:title => @right_cell_text)
+        else
+          # Ancestry parents breadcrumbs (only in services)
+          breadcrumbs.concat(ancestry_parents(options[:ancestry], options[:record_info], options[:record_title])) if options[:ancestry]
 
-        # Append breadcrumbs created from the tree (eg "All policies > Red Hat policies > Policy 1")
-        breadcrumbs_from_tree = build_breadcrumbs_from_tree
-        breadcrumbs.concat(breadcrumbs_from_tree)
+          # Try to create treenode from the x_node
+          entity, entity_id = TreeBuilder.extract_node_model_and_id(x_node)
+          record = entity.try(:constantize).try(:find, entity_id)
+          title = TreeNode.new(record).text if record
 
-        if (options[:include_record] || breadcrumbs_from_tree.blank?) && options[:record_info].present? && options[:record_info][options[:record_title]] &&
-           !(@tagitems || @politems || @ownershipitems || @retireitems)
-          # Append breadcrumb created from record_info if included or if result from tree was empty (eg filters page)
+          # otherwise get the text from different sources
+          unless title
+            # Item in the tree + fallbacks
+            title = x_node_text || options[:record_info][options[:record_title]] || @title_for_breadcrumbs
 
-          # Ancestry parents breadcrumbs
-          breadcrumbs.concat(ancestry_parents(options[:ancestry], options[:record_info], options[:record_title])) if options[:ancestry]
+            # Initial loading of the selected node when coming from different page
+            title = build_breadcrumbs_from_tree.last.try(:[], :title) if @trees.present?
```

Comment:
I guess you don't really need the whole path to the breadcrumbs, just find the last element. So the `build_breadcrumbS_from_tree` method can simplified a lot and maybe even singularized to `build_breadcrumb_from_tree`.